porch/pkg/engine/engine.go

```diff
@@ -420,61 +423,91 @@ func (cad *cadEngine) ListFunctions(ctx context.Context, repositoryObj *configap
 	task              *api.Task
 	cad               CaDEngine
 	referenceResolver ReferenceResolver
 	namespace         string
+	pkgName           string
 }
 
 func (m *updatePackageMutation) Apply(ctx context.Context, resources repository.PackageResources) (repository.PackageResources, *api.Task, error) {
 	ctx, span := tracer.Start(ctx, "updatePackageMutation::Apply", trace.WithAttributes())
 	defer span.End()
 
-	// TODO: load directly from source repository
-	dir, err := ioutil.TempDir("", "kpt-pkg-update-*")
+	currUpstreamPkgRef, err := m.currUpstream()
 	if err != nil {
 		return repository.PackageResources{}, nil, err
 	}
-	defer os.RemoveAll(dir)
 
-	if err := writeResourcesToDirectory(dir, resources); err != nil {
-		return repository.PackageResources{}, nil, err
+	targetUpstream := m.task.Clone.Upstream
+	if targetUpstream.Type == api.RepositoryTypeGit || targetUpstream.Type == api.RepositoryTypeOCI {
+		return repository.PackageResources{}, nil, fmt.Errorf("update is not supported for non-porch upstream packages")
 	}
 
-	var ref string
-	var packageDir string
-	upstream := m.task.Clone.Upstream
-	switch upstream.Type {
-	case api.RepositoryTypeGit:
-		ref = upstream.Git.Ref
-		packageName := filepath.Base(m.task.Clone.Upstream.Git.Directory)
-		packageName = strings.TrimPrefix(packageName, ".git")
-		packageDir = filepath.Join(dir, packageName)
-	case api.RepositoryTypeOCI:
-		return repository.PackageResources{}, nil, fmt.Errorf("update is not supported for oci packages")
-	default:
-		revision, err := (&PackageFetcher{
-			cad:               m.cad,
-			referenceResolver: m.referenceResolver,
-		}).FetchRevision(ctx, upstream.UpstreamRef, m.namespace)
-		if err != nil {
-			return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package revisions %s", upstream.UpstreamRef.Name)
-		}
-		revisionKey := revision.Key()
-		ref = fmt.Sprintf("%s/%s", revisionKey.Package, revisionKey.Revision)
-		packageDir = dir
+	originalResources, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchResources(ctx, currUpstreamPkgRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package %s with ref %+v",
+			m.pkgName, *currUpstreamPkgRef)
 	}
 
-	if err := kpt.PkgUpdate(ctx, ref, packageDir, kpt.PkgUpdateOpts{}); err != nil {
-		return repository.PackageResources{}, nil, err
+	upstreamRevision, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchRevision(ctx, targetUpstream.UpstreamRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching revision for target upstream %s", targetUpstream.UpstreamRef.Name)
+	}
+	upstreamResources, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchResources(ctx, targetUpstream.UpstreamRef, m.namespace)
```

Comment:
Might be easier to just fetch the resources from the upstreamRevision here with `upstreamRevision.GetResources(ctx)`.

Reply:
Good pt. Done.